### PR TITLE
Filter pods by labelSelector

### PR DESCRIPTION
**Pull request by lixiang233**
Fixes: #195 

TODO:
- [x] API change
- [x] new labelSelector option
- [x] Add option to strategies
- [x] README and e2e-tests

/kind feature

### CODE REVIEW

File: pkg/descheduler/strategies/topologyspreadconstraint.go [hunk 1]
```diff
@@ -110,13 +120,12 @@ func RemovePodsViolatingTopologySpreadConstraint(
 	podsForEviction := make(map[*v1.Pod]struct{})
 	// 1. for each namespace...
 	for _, namespace := range namespaces.Items {
-		if (len(includedNamespaces) > 0 && !includedNamespaces.Has(namespace.Name)) ||
-			(len(excludedNamespaces) > 0 && excludedNamespaces.Has(namespace.Name)) {
```
Comment (lixiang233) [inline pkg/descheduler/strategies/topologyspreadconstraint.go:114]: modify this to pass cyclomatic complexity check.

**Comment by seanmalloy** [comment]
@lixiang233 is this PR ready for review?

**Comment by lixiang233** [comment]
> @lixiang233 is this PR ready for review?

yes, PTAL when you have time.

File: test/e2e/e2e_test.go [hunk 2]
```diff
@@ -517,6 +526,109 @@ func testPriority(t *testing.T, isPriorityClass bool) {
...
+	t.Logf("Waiting 10s")
+	time.Sleep(10 * time.Second)
+	// check if all pods without target label are not evicted
+	podListReserve, err = clientSet.CoreV1().Pods(rcReserve.Namespace).List(
+		ctx, metav1.ListOptions{LabelSelector: labels.SelectorFromSet(rcReserve.Spec.Template.Labels).String()})
+	if err != nil {
+		t.Fatalf("Unable to list pods after running strategy: %v", err)
+	}
+
+	excludePodNames := getPodNames(podListReserve.Items)
```
Comment (ingvagabund) [inline test/e2e/e2e_test.go:591]: s/excludePodNames/reservedPodNames

File: test/e2e/e2e_test.go [hunk 3]
```diff
@@ -517,6 +526,109 @@ func testPriority(t *testing.T, isPriorityClass bool) {
...
+
+	//check if all selected pods are evicted
+	if err := wait.PollImmediate(time.Second, 20*time.Second, func() (bool, error) {
+		podListEvict, err := clientSet.CoreV1().Pods(rcEvict.Namespace).List(
+			ctx, metav1.ListOptions{LabelSelector: labels.SelectorFromSet(rcEvict.Spec.Template.Labels).String()})
+		if err != nil {
+			return false, nil
+		}
+
+		includePodNames := getPodNames(podListEvict.Items)
```
Comment (ingvagabund) [inline test/e2e/e2e_test.go:608]: s/includePodNames/newPodNames/

**Review by ingvagabund — COMMENTED** [review]
Allowing to set label selector in `RemoveDuplicates` and `RemovePodsViolatingTopologySpreadConstraint` improperly might result in the opposite of what those strategies wants to achieve. The skew of pods between topology domains might actually increase.

File: README.md [hunk 4]
```diff
@@ -498,6 +498,40 @@ strategies:
...
+The following strategies can configure a [standard kubernetes labelSelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#labelselector-v1-meta)
+to filter pods by their labels:
+
+* `PodLifeTime`
+* `RemovePodsHavingTooManyRestarts`
+* `RemovePodsViolatingNodeTaints`
+* `RemovePodsViolatingNodeAffinity`
+* `RemovePodsViolatingInterPodAntiAffinity`
+* `RemoveDuplicates`
+* `RemovePodsViolatingTopologySpreadConstraint`
```
Comment (damemi) [inline README.md:512]: Could you also update the parameters tables in the README for each of these strategies? (like https://github.com/kubernetes-sigs/descheduler/commit/11b9829885ea8582c5fdd773f179d6bda1ed3d24)

**Comment by lixiang233** [comment]
> Allowing to set label selector in `RemoveDuplicates` and `RemovePodsViolatingTopologySpreadConstraint` improperly might result in the opposite of what those strategies wants to achieve. The skew of pods between topology domains might actually increase.

You are right, so should we mention this in README to tell users the risk of using this parameter or just disable it in these strategies?

**Comment by seanmalloy** [comment]
> > Allowing to set label selector in `RemoveDuplicates` and `RemovePodsViolatingTopologySpreadConstraint` improperly might result in the opposite of what those strategies wants to achieve. The skew of pods between topology domains might actually increase.
> 
> You are right, so should we mention this in README to tell users the risk of using this parameter or just disable it in these strategies?

Disable it in these strategies.

**Comment by ingvagabund** [comment]
To disable. We don't want to allow users to configure a strategy the way it malfunctions.

**Comment by lixiang233** [comment]
Disabled. PTAL @ingvagabund  @seanmalloy

**Comment by ingvagabund** [comment]
Worth changing `Strategies: Add labelSelector to all strategies except LowNodeUtilization` commit message.

**Comment by lixiang233** [comment]
> Worth changing `Strategies: Add labelSelector to all strategies except LowNodeUtilization` commit message.

updated.

**Comment by damemi** [comment]
I think this looks all good to me, anyone else have any suggestions?
/approve

**Review by seanmalloy — APPROVED** [review]

**Comment by seanmalloy** [comment]
It looks good to me. I did not find anything that needs changing. I'll give @ingvagabund some time to review just in case he want to take one more look.

Also FYI @killianmuldoon the long awaited label filtering feature that you suggested has been implemented. Feel free to review if you have time. Thanks!

/cc @killianmuldoon

**Comment by ingvagabund** [comment]
/lgtm

Thank you @lixiang233!!!

**Pull request merged**